torch/standalone/header_only/core/DeviceType.h

```diff
@@ -0,0 +1,288 @@
+  COMPILE_TIME_MAX_DEVICE_TYPES = 21,
+};
+
+constexpr DeviceType kCPU = DeviceType::CPU;
+constexpr DeviceType kCUDA = DeviceType::CUDA;
+constexpr DeviceType kHIP = DeviceType::HIP;
+constexpr DeviceType kFPGA = DeviceType::FPGA;
+constexpr DeviceType kMAIA = DeviceType::MAIA;
+constexpr DeviceType kXLA = DeviceType::XLA;
+constexpr DeviceType kMPS = DeviceType::MPS;
+constexpr DeviceType kMeta = DeviceType::Meta;
+constexpr DeviceType kVulkan = DeviceType::Vulkan;
+constexpr DeviceType kMetal = DeviceType::Metal;
+constexpr DeviceType kXPU = DeviceType::XPU;
+constexpr DeviceType kHPU = DeviceType::HPU;
+constexpr DeviceType kVE = DeviceType::VE;
+constexpr DeviceType kLazy = DeviceType::Lazy;
+constexpr DeviceType kIPU = DeviceType::IPU;
+constexpr DeviceType kMTIA = DeviceType::MTIA;
+constexpr DeviceType kPrivateUse1 = DeviceType::PrivateUse1;
+
+// define explicit int constant
+constexpr int COMPILE_TIME_MAX_DEVICE_TYPES =
+    static_cast<int>(DeviceType::COMPILE_TIME_MAX_DEVICE_TYPES);
+
+static_assert(
+    COMPILE_TIME_MAX_DEVICE_TYPES <= 21,
+    "Hey!  You seem to be adding a lot of new DeviceTypes.  The intent was "
+    "for this constant to reflect the actual number of DeviceTypes we support "
+    "in PyTorch; it's important that this number is not too large as we "
+    "use this to allocate stack arrays in some places in our code.  If you "
+    "are indeed just adding the 20th device type, feel free to change "
+    "the check to 32; but if you are adding some sort of extensible device "
+    "types registration, please be aware that you are affecting code that "
+    "this number is small.  Try auditing uses of this constant.");
+
+// We use both a mutex and an atomic here because:
+// (1) Mutex is needed during writing:
+//     We need to first check the value and potentially error,
+//     before setting the value (without any one else racing in the middle).
+//     It's also totally fine for this to be slow, since it happens exactly once
+//     at import time.
+// (2) Atomic is needed during reading:
+//     Whenever a user prints a privateuse1 device name, they need to read this
+//     variable. Although unlikely, we'll data race if someone else is trying to
+//     set this variable at the same time that another thread is print the
+//     device name. We could re-use the same mutex, but reading the atomic will
+//     be much faster.
+//
+// Note: Can not directly use inline to declar a global variable here, because
+//       it doesn't work for fbobjc build, error: declaration requires a global
+//       destructor C10_EXPORT is also needed to make sure the variables are
+//       visible to all files.
+inline C10_EXPORT std::atomic<bool>& privateuse1_backend_name_set() {
+  static std::atomic<bool> privateuse1_backend_name_set_(false);
+  return privateuse1_backend_name_set_;
+}
+
+inline C10_EXPORT std::string& privateuse1_backend_name() {
+  static std::string privateuse1_backend_name_;
```

Comment:
Aren't static variables in headers a big no-no as it makes it local per-compilation unit which is very much not what we want here?

Reply:
`inline` removes your concern. If we declared the function as `static` or `inline static`, then yes, it won't work as you mentioned. https://stackoverflow.com/questions/185624/what-happens-to-static-variables-in-inline-functions

Reply:
This explanation is quite confusing to me. My read is that it will work within a single compilation unit but not across compilation units?

In particular, if I have this function inlined in libtorch.so and another version inlined in libuser.so. You're saying that the compiler will figure out a way to make them shared between the two? Which one of the two .so will actually host the shared variable? I'm quite confused how this can work.

Reply:
I think it's not the compiler but the linker which will only keep one copy of the symbol. The only alternative I can think of is to keep some .cpp files under the standalone directory. Let me know what do you think of that approach.

Reply:
As an example, with my change, nm shows `privateuse1_backend_name` is of type `u`:

           "u" The symbol is a unique global symbol.  This is a GNU extension to the standard set of ELF symbol bindings.  For such a symbol the dynamic linker will make sure that in the
               entire process there is just one symbol with this name and type in use.

Reply:
Ho that's pretty cool!
Given it's a GNU extension, will this work on clang and msvc as well?

Reply:
> Given it's a GNU extension, 

As per the stack overflow answer earlier in this thread, the standard requires that "A static local variable in an extern inline function always refers to the same object." A particular platform choosing to use platform-specific things to implement this requirement does not change that the requirement is coming from the standard.